readme.MD
#Скрипт фотохостинга "Akina"

- Название: Akina
- Назначение: фотохостинг
- Год выпуска: 2011-2015
- Лицензия: GNU GPL v2
- Язык Интерфейса: Руский
- Платформа/ОС: ALL (но желательно Unix-подобные)
- Сайт разработчиков: akina-photohost.org
- Написано на: PHP, JavaScript
- Размер: 270 KB

##Описание
Бесплатный, простой и многофункциональный скрипт фотохостинга. Не требует много ресурсов, не использует базы данных. Поддерживает мультизагрузуку и уменьшение изображений.

##Системные требования
* PHP 5.0 и выше
* GD2 и выше
* cURL

##Пользовательские возможности
* загрузка изображений jpg, gif, png, bmp
* мультизагрузка с локального компьютера и с удаленного сервера (допустимо одновременно)
* единое поле для загрузки изображений с удаленных серверов
* опционально уменьшение изображений
* опциональное создание превью
* опциональное создание надписи на превью (размеры изображения / собственный текст)
* генерация ссылок на страницу просмотра, прямой ссылки на изображение, BB и html-кодов.
* генерация кодов с превью.
* при мультизагрузке генерируются общие BB и html-коды всех загруженных изображений

##Остальные возможности скрипта
* возможность установки в поддиректорию
* поддержка шаблонов
* страница просмотра загруженного изображения
* возможность ограничения предельно допустимого размера и разрешения загружаемого изображения
* изменение качества для jpg-изображений
* отображение лимита количества отдновременно загружаемых изображений, разрешенных форматов изображений и их размеров
* проверка загружаемых объектов по mime-type
* разбивка загрузки изображений по каталогам год-месяц/день
* вывод статистики количества загруженных на фотохостинг изображений за все время, за последние сутки и занимаемый объем

##Известные недоработки
https://github.com/vla7/akina/issues

##Планы
https://github.com/vla7/akina/issues

##УСТАНОВКА
1. Залить все файлы на хостинг
2. Установить владельцем файлов www-data (или ваш владелец процесса Apache) и права 755 для каталогов /img/, /thumbs/, /working/, /working/thumbs/ и файла /working/cachefile.dat
3. Пройтись по конфигу (файл config.php) и настроить все под себя. Например, выбрать шаблон из доступных в папке templates правкой переменной $config['template_name']
По умолчанию включен шаблон whatsyoursolution
4. Добавление статических страниц.
Файлы, в формате html можно размещать в папке с вашим шаблоном. Название должно быть в формате blablabla.static.tpl (".static.tpl" в конце - обязательно).
После размещения файла эта страница будет доступна по адресу http://yoursite.com/?p=blablabla (без ".static.tpl" в конце)

##ОБНОВЛЕНИЕ
###Переход с версии 1.0.8 и ниже, на версии 1.0.9 и выше
1. Залить на хостинг файл __checker.php
2. Запустить его http://yoursite.com/__checker.php
3. Внести изменения в файл config.php из дистрибутива 1.0.9 в поля подсвеченные красным цветом, (на локальном диске)
4. Если есть ваши собственные шаблоны или вы вносили в них изменения, сохранить их на локальный диск.
5. Залить все файлы на хостинг
6. Удалить файл __checker.php (он уже не нужен)
7. Пройтись по конфигу (файл config.php) - настроить остальное под себя.
При необходимости внести изменения в шаблоны.



CHANGELOG.txt
LEGEND
------
  [!] New ???
  [+] New feature
  [*] Enhancement
  [-] Bug fix

v1.0.9i (r79 / github) Jun 28, 2015
------
  [+] BMP всегда преобразуется к PNG.
  [*] Оптимизация кода.
  [+] Вся работа с графикой переведена на GD Enhancer.
  [*] Шрифт используемый в нанесении текста на превью myfont.phpfont заменен на OpenType Bender (bender.otf).
      Можно менять шрифт на любой в формате TrueType или OpenType.

v1.0.9h (r78 / github) Jun 26, 2015
------
  [+] Добавлен скрипт __checker.php сравнения настроек config.php с параметрами "по умолчанию" для перехода на новую версию.
  [*] Добавлено описание процесса перехода на новую версию в readme.txt.

v1.0.9h (r77 / github) Jun 26, 2015
------
  [*] Обновлена библиотека GD Enhancer до версии 4.01 (https://github.com/coldume/gd-enhancer)
  [*] Добавлена проверка подключения файлов библиотеки GD Enhancer.
  [+] Оптимизация кода. Убрано обращение к функциям finfo_open и mime_content_type.
  [-] Issue 85 : "finfo_open" и "mime_content_type"

v1.0.9g (github) Jun 25, 2015
v1.0.9g (r76) Jun 18, 2015
------
  [-] Исправлены ошибка "Изображение не загружено. Ошибка: Неверный MIME-тип изображения, допускаются JPEG, GIF, PNG, BMP. Вы пытались залить".
      Если некорректно установлены права на каталоги.
  [+] Добавлена проверка прав на каталоги.
  [*] Добавлен в config.php флаг debug = [true|false] отладка, показывать ошибки PHP включено/выключено
  [-] Исправлены ошибки в коде, не влияющие на работу, но дающие сообщения в протокол WEB сервера.
  [*] Добавлена проверка подключения файлов config.php и functions.php.

v1.0.9f (r75) Oct 10, 2014
------
  [+] Исправлена ошибка шаблона 'BlueStyle', не выделялись поля с кодом по клику.

v1.0.9f (r74) Sep 10, 2014
------
  [+] Новый шаблон 'BlueStyle' Дизайн: reflex , Кодинг: TerVel

v1.0.9f (r73) Sep 10, 2014
------
  [-] Issue 2 : Поддержка уменьшения анимированного gif
  [*] Замена библиотеки GIF_eXG (https://github.com/Yuriy-Khomenko/GIF_eXG) на GD Enhancer (http://www.gdenhancer.com/)
  [-] Исправлена ошибка/добавлена поддержка уменьшения анимированных GIF с прозрачным фоном.
  [-] Подправлен config.php , убран дубль 'bmp' .

v1.0.9e (r71,r72) Sep 9, 2014
------
  [-] Issue 2 : Поддержка уменьшения анимированного gif
  [-] CHANGELOG.txt сконвертирован в UTF-8 для корректного отображения в SVN

v1.0.9d (r70) Sep 5, 2014
------
  [-] Исправлена ошибка/добавлена поддержка создания превью для BMP формата.

v1.0.9d (r68,r69) Sep 4, 2014
------
  [-] Issue 1 
  [+] Добавлена поддержка BMP формата.

v1.0.9c (r65,r66,r67) Aug 26, 2014
------
  [-] Issue 38
  [+] Добавлен в config.php флаг: выводить дату/время загрузки изображения на "Странице просмотра"

v1.0.9b Aug 22, 2014
------
  [-] Issue 39
  [+] Настройки текста в тэгах title,h1,h2 титульной страницы шаблона вынесены в config.php
  [-] Issue 73
  [+] Добавлен в config.php флаг: добавлять к коду "изображения с превью" ссылку на "Страницу просмотра"

v1.0.9a Aug 20, 2014
------
  [-] Issue 79
  [-] Issue 84
  [-] Issue 86
  [+] Добавлены в config.php флаги: "по умолчанию" для форм создания превью, уменьшение размера картинки
  [+] Добавлены в config.php: параметры в формах создания превью, уменьшение размера картинки
  [-] Issue 87
  [-] Issue 75, 76, 83
  [+] Настройка файла статистики вынесена в файл config.php , путь по умолчанию теперь 'site_dir'/working/cachefile.dat (раньше был просто 'site_dir'/cache)
  [*] В архив включены каталоги /img/ , /thumbs/ , /working/ , /working/thumbs/
  [*] в HTML код добавлен параметр открытия в новом окне (Issue 73 частично)
  [-] Issue 74 : 	Ошибка с часовым поясом
  [*] Добавлено в config.php : настройка временной зоны date_default_timezone_set("Europe/Minsk"); Возможные параметры: http://php.net/manual/en/timezones.php

v1.0.8 Sep 2, 2013
------
  [-] критическая уязвимость во всех версиях движка

v1.0 Sep 6, 2011
------
  [*] Cтабильная версия 1.0


__checker.php
<?php
define('akina', 'photohost', true);

// расцветка для подсветки элементов
$check_good="#33AA33";
$check_bad="#FF3333";
$check_empty="#666666";
$check_element="#FF7F00";

//загружаем предустановки
if(!include_once('config.php')) die('Can\'t find config.php');
?>

<head>
	<title>Akina. Проверка перехода на новую версию. Checking move to a new version</title>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<style>
body {
    font-family: Courier, 'Courier New', 'Times New Roman', Times, serif;
    font-size: 11pt;
}
</style>
</head>
<body >
<?php 

function new_config()
{
/* --------------------------------------------------------------------------
 Далее содержится файл конфигурации config.php поставляемый по "умолчанию" 
 По состоянию на v1.0.9h (r77) Jun 26, 2015 
----------------------------------------------------------------------------- */

//////////////////////////////////////параметры изображений//////////////////////////////////////
$config['max_size_mb']=5;
$config['max_size_byte']=$config['max_size_mb']*1048576; // (bytes)
$config['max_height']=5000;
$config['max_width']=5000;
$config['view_one_width']="auto"; //ширина картинки на странице просмотра. картинка одна, большая
$config['view_multi_width']="auto"; //ширина картинок на странице просмотра при мультизагрузке

$config['quality']=100;

$config['mimes']=array('image/gif', 'image/pjpeg', 'image/jpeg', 'image/png', 'image/bmp', 'image/x-ms-bmp');
//расширения которые можно загружать (jpeg и jpe - это jpg)
$config['extensions']=array(
							'gif',
							'jpg',
							'png',
							'bmp',
);

$config['auto_resize']=0;  //Уменьшить изображение, по умолчанию форма: 0 - выключена, 1 - включена
$config['width_resize_elements']=1024; //уменьшать изображения по ширине, по умолчанию в форме
//$config['height_resize_elements']=768; //уменьшать изображения по высоте, по умолчанию в форме
// для исключения искажения изображения добавляется только один параметр, по ширине - имеет приоритет.

$config['auto_preview']=0;  //Создать превью, по умолчанию форма: 0 - выключена, 1 - включена
$config['width_preview_elements']=240; //превью по ширине, по умолчанию в форме
//$config['height_preview_elements']=180; //превью по высоте, по умолчанию в форме
// для исключения искажения изображения добавляется только один параметр, по ширине - имеет приоритет.

//////////////////////////////////////абсолютные пути//////////////////////////////////////

$config['site_dir']=getcwd();
$config['uploaddir']=$config['site_dir'].'/img/';
$config['thumbdir']=$config['site_dir'].'/thumbs/';
$config['working_dir']=$config['site_dir'].'/working/';
$config['working_thumb_dir']=$config['working_dir'].'thumbs/';

//////////////////////////////////////URL//////////////////////////////////////

preg_match('/\/(.*\/)index.php/', $_SERVER['PHP_SELF'], $out);
$folder = isset($out[1]) ? $out[1]:'';
$config['site_url']='http://'.$_SERVER['HTTP_HOST'].'/'.$folder;
$config['thumbs_url']=$config['site_url'].'thumbs/';
$config['img_url']=$config['site_url'].'img/';

//////////////////////////////////////Шаблон//////////////////////////////////////

//доступные шаблоны 'bluestyle', 'graphene' , 'simple', 'whatsyoursolution'
$config['template_name']='whatsyoursolution'; 
$config['template_path']=$config['site_dir'].'/templates/'.$config['template_name'];
$config['template_url']=$config['site_url'].'templates/'.$config['template_name'];

$config['site_title']='Хостинг картинок AKINA'; //Title страницы, возможно динамическое обновление 
$config['site_header_h1']='Фотохостинг Akina'; //Текст в заголовке сайта тэг H1, возможно динамическое обновление 
$config['site_header_h2']='Хостинг картинок'; //Текст в заголовке сайта тэг H2, возможно динамическое обновление

$config['view_page']=1; //добавлять к коду изображения с превью ссылку на "Страницу просмотра" 0 - выключено, 1 - включено
//если выключено всегда будет ссылка вести прямо на оригинальную картинку 

$config['show_upload_date']=1; //показывать дату/время загрузки изображения на "Странице просмотра" 0 - выключено, 1 - включено

//////////////////////////////////////CURL//////////////////////////////////////

$config['curl_timeout'] = 120;//таймаут для curl
$config['curl_user_agent']='User-Agent: Opera/9.80 (X11; Linux i686; U; ru) Presto/2.9.168 Version/11.50';
$config['curl_headers']=array(
'GET / HTTP/1.1',
'Accept: text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1',
'Accept-Language: ru,ru-RU;q=0.9,en;q=0.8',
'Accept-Charset: iso-8859-1, utf-8, utf-16, *;q=0.1',
'Accept-Encoding: deflate, gzip, x-gzip, identity, *;q=0',
'Cookie: cookies_enabled=1;',
'Cache-Control: no-cache',
'Connection: Keep-Alive, TE',
'TE: deflate, gzip, chunked, identity, trailers'
);

//////////////////////////////////////Даты - для создания каталогов//////////////////////////////////////

$config['current_month']=date ('Y-m');
$config['current_day']=date ('d');
$config['current_path']=$config['current_month'].'/'.$config['current_day'];
$debug = true; // [true|false] отладка, показывать ошибки PHP включено/выключено

//////////////////////////////////////Прочее//////////////////////////////////////

$config['random_str_quantity']=25;
$config['site_work']=true; // [true|false] сайт работает да/нет
$config['cache_time']=60*60; //в секундах, 1 час.
$config['cachefile']=$config['working_dir']."/cachefile.dat"; //файл статистики "Изображений на фотохостинге: х; занимают х.х Kb; за сутки загружено: х"

/* --------------------------------------------------------------------------
Конец файла конфигурации config.php поставляемый по "умолчанию"
----------------------------------------------------------------------------- */

return ($config);
}

$new_config = new_config();

echo "<p>";
// расцветка для подсветки элементов
echo "Расцветка элементов после проверки:<br />\n";
echo "<ul>";
echo "<li><font color='$check_element'> Название элемента из конфигурации.</font><br />\n";
echo "<li><font color='$check_good'> Элемент <b>совпадает</b> с вашей конфигурацией.</font><br />\n";
echo "<li><font color='$check_bad'> Нужна корректировка параметра. Элемент есть и <b>не совпадает</b> с вашей конфигурацинй.</font><br />\n";
echo "<li><font color='$check_empty'> Элемент <b>отсутствует</b> в вашей конфигурацинй.</font><br />\n";
echo "</ul>";
echo "</p>";

//echo "<pre>"; print_r($new_config); echo "</pre>";

echo "<hr><br>";

$level=0;
function check_array($array,$arr_key)
{
	GLOBAL $config,$new_config,$level;
	GLOBAL $check_good,$check_bad,$check_empty,$check_element;
	foreach ($array as $key => $value)
	{
		$check_b="<font color='$check_empty'>";
		$check_e="</font>";
	
  	if (!is_array($value))
  	{
      if (!empty($arr_key))
      {
				if (in_array($value, $config[$arr_key]))
	 			{
				  $check_b="<font color='$check_good'>";
				  $check_e="</font>";
				}
			}
			else
			if (isset($config[$key]))
			{
				if ($config[$key]===$new_config[$key])
				{
				  $check_b="<font color='$check_good'>";
				  $check_e="</font>";
				}
				else
				{
				  $check_b="<font color='$check_bad'>";
				  $check_e="</font>";
				}
			}

		  for ($i=0;$i < $level+1 ;$i++)  echo "&nbsp;&nbsp;&nbsp;&nbsp;";
		  
 		  if ($level) $ccf=''; else  $ccf="<font color='$check_element'>\$config</font>";
			echo $ccf."['<font color='$check_element'>".$key."</font>'] => ".$check_b.$value."&nbsp".$check_e."<br />\n";
		}
		else
		{
			if (array_key_exists($key, $config))
			{
			  $check_b="<font color='#00CC00'>";
			  $check_e="</font>";
			}

		  for ($i=0;$i < $level+1 ;$i++)  echo "&nbsp;&nbsp;&nbsp;&nbsp;";
			echo "<font color='$check_element'>\$config</font>['<font color='$check_element'>".$key."</font>'] => ".$check_b."Array".$check_e."<br />\n";

		  for ($i=0;$i < $level+1 ;$i++)  echo "&nbsp;&nbsp;&nbsp;&nbsp;";
		  echo "&nbsp;&nbsp;(<br />\n";
			$level++;
			check_array($value,$key);
			$level--;
		  for ($i=0;$i < $level+1 ;$i++)  echo "&nbsp;&nbsp;&nbsp;&nbsp;";
		  echo "&nbsp;&nbsp;)<br />\n";
		}
	}
    
}

check_array($new_config,NULL);

if ((!isset($config['cachefile'])) and (file_exists($config['site_dir']."/cache" )) )
echo "<font color='$check_bad'>** Нужна корректировка параметра</font> <font color='$check_element'>\$config</font>['<font color='$check_element'>cachefile</font>']=<font color='$check_element'>\$config</font>['<font color='$check_element'>site_dir</font>'].'/cache'</font>\n";


?>
<br><hr><br>

</body>
</html>
